type-definitions/most.d.ts

```diff
@@ -0,0 +1,144 @@
+declare type SeedValue<S, V> = { seed: S, value: V };
+declare type TimeValue<V>    = { time: number, value: V };
+
+declare interface Promise<A> {}
+declare interface Generator<A, B, C> {}
+declare interface Iterable<A> {}
+
+declare type CreateGenerator<A> = (...args: Array<any>) => Generator<A|Promise<A>, any, any>;
+
+export interface Stream<A> {
+  reduce<B>(f: (b: B, a: A) => B, b: B): Promise<B>;
+  observe(f: (a: A) => any): Promise<any>;
+  forEach(f: (a: A) => any): Promise<any>;
+  drain(): Promise<any>;
+
+  constant<B>(b: B): Stream<B>;
+  map<B>(f: (a: A) => B): Stream<B>;
+  tap(f: (a: A) => any): Stream<A>;
+  chain<B>(f: (a: A) => Stream<B>): Stream<B>;
+  flatMap<B>(f: (a: A) => Stream<B>): Stream<B>;
+
+  // Note:  Without higher-kinded types, the types for these
+  // cannot be written properly
+  ap<B, C>(bs: Stream<B>): Stream<C>;
+  join<B>(): Stream<B>;
+  switch<B>(): Stream<B>;
+  switchLatest<B>(): Stream<B>;
+
+  continueWith(f: (a: any) => Stream<A>): Stream<A>;
+  concatMap<B>(f: (a: A) => Stream<B>): Stream<B>;
+  mergeConcurrently<B>(concurrency: number): Stream<B>;
+  merge(...ss: Array<Stream<A>>): Stream<A>;
+  combine<B>(f: (a: A, ...args: Array<any>) => B, ...ss: Array<Stream<any>>): Stream<B>;
+
+  scan<B>(f: (b: B, a: A) => B, b: B): Stream<B>;
+  loop<S, B>(f: (seed: S, a: A) => SeedValue<S, B>, seed: S): Stream<B>;
+
+  cycle(): Stream<A>;
+  concat(s2: Stream<A>): Stream<A>;
+  startWith(a: A): Stream<A>;
+
+  filter(p: (a: A) => boolean): Stream<A>;
+  skipRepeats(): Stream<A>;
+  skipRepeatsWith(eq: (a1: A, a2: A) => boolean): Stream<A>;
+
+  take(n: number): Stream<A>;
+  skip(n: number): Stream<A>;
+  takeWhile(p: (a: A) => boolean): Stream<A>;
+  skipWhile(p: (a: A) => boolean): Stream<A>;
+  slice(start: number, end: number): Stream<A>;
+
+  until(signal: Stream<any>): Stream<A>;
+  takeUntil(signal: Stream<any>): Stream<A>;
+  since(signal: Stream<any>): Stream<A>;
+  skipUntil(signal: Stream<any>): Stream<A>;
+  during(timeWindow: Stream<Stream<any>>): Stream<A>;
+  throttle(period: number): Stream<A>;
+  debounce(period: number): Stream<A>;
+
+  timestamp(): Stream<TimeValue<A>>;
+  delay(dt: number): Stream<A>;
+
+  // Note:  Without higher-kinded types, this type cannot be written properly
+  await<B>(): Stream<B>;
+
+  sample<B>(f: (...as: any[]) => B, ...ss: Array<Stream<any>>): Stream<B>;
+  sampleWith<A>(sampler: Stream<any>): Stream<A>;
+
+  zip<A>(f: (a: A, ...as: any[]) => A, ...ss: Array<Stream<any>>): Stream<A>;
+
+  recoverWith<B>(p: (a: B) => A): Stream<A>;
+  multicast<A>(): Stream<A>;
+}
+
+export function create<A>(f: (add: Function, end: Function, error: Function) => void): Stream<A>;
```

Comment:
I think we can make `create` a bit tighter.  My TS syntax is probably wrong, but hopefully gives the idea:

```ts
export function create<A>(f: (add: (a:A) => any, end: (x:any) => any, error: (e:Error) => any) => (?() => ?Promise<any>)): Stream<A>;
```

The most important thing, IMHO, is tying `add`'s parameter type, to the type parameter of the returned `Stream<A>`.  Also, the function passed to `create` can optionally return a cleanup function, which can optionally return a Promise to represent an async cleanup process.

We might be able to try to incorporate the stream end value as one of the `Stream` type parameters, e.g. `Stream<A, Z>`, but I don't think we need to go there for the first version of this :)

Did that make sense?